src/database/DataTypes/Transaction.js

```diff
@@ -157,21 +210,62 @@ export class Transaction extends Realm.Object {
   confirm(database) {
     if (this.isConfirmed) throw new Error('Cannot confirm as transaction is already confirmed');
     if (this.isFinalised) throw new Error('Cannot confirm as transaction is already finalised');
-    this.items.forEach((transactionItem) => {
-      transactionItem.batches.forEach((transactionBatch) => {
-        const itemBatch = transactionBatch.itemBatch;
-        const newNumberOfPacks = this.isIncoming ?
-          itemBatch.numberOfPacks + transactionBatch.numberOfPacks :
-          itemBatch.numberOfPacks - transactionBatch.numberOfPacks;
-        itemBatch.packSize = transactionBatch.packSize;
-        itemBatch.numberOfPacks = newNumberOfPacks;
-        itemBatch.expiryDate = transactionBatch.expiryDate;
-        itemBatch.batch = transactionBatch.batch;
-        itemBatch.costPrice = transactionBatch.costPrice;
-        itemBatch.sellPrice = transactionBatch.sellPrice;
-        database.save('ItemBatch', itemBatch);
-      });
+
+    this.pruneRedundantItems(database);
+    const isIncomingInvoice = this.isIncoming;
+    this.transactionBatches(database).forEach(transactionBatch => {
+      const { itemBatch,
+              batch,
+              packSize,
+              numberOfPacks,
+              expiryDate,
+              costPrice,
+              sellPrice,
+             } = transactionBatch;
+
+      const newNumberOfPacks = isIncomingInvoice
+          ? itemBatch.numberOfPacks + numberOfPacks
+          : itemBatch.numberOfPacks - numberOfPacks;
+      itemBatch.packSize = packSize;
+      itemBatch.numberOfPacks = newNumberOfPacks;
+      itemBatch.expiryDate = expiryDate;
+      itemBatch.batch = batch;
+      itemBatch.costPrice = costPrice;
+      itemBatch.sellPrice = sellPrice;
+      database.save('ItemBatch', itemBatch);
+    });
+
+    this.confirmDate = new Date();
+    this.status = 'confirmed';
+  }
+
+  /**
+   * Confirm this transaction, generating the associated item batches, linking them
+   * to their items, and setting the status to confirmed.
+   * @param  {Realm}  database The app wide local database
+   * @return {none}
+   */
+  confirmExternalSI(database) {
+    if (this.isConfirmed) throw new Error('Cannot confirm as transaction is already confirmed');
+    if (this.isFinalised) throw new Error('Cannot confirm as transaction is already finalised');
+
+    this.pruneRedundantBatches(database);
```

Comment:
I though there will be too many conditionals if we combine these two together.  Let's see how it looks though, will commit as advised.